ocrd_network/ocrd_network/processing_server.py

```diff
@@ -817,3 +859,51 @@ async def get_workflow_info(self, workflow_job_id) -> Dict:
                     "page_id": job.page_id,
                 })
         return res
+
+    async def upload_workflow(self, workflow: UploadFile) -> Dict:
+        """ Store a script for a workflow in the database
+        """
+        workflow_id = generate_id()
+        content = (await workflow.read()).decode("utf-8")
+        if not validate_workflow(content):
+            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                                detail="Provided workflow script is invalid")
+
+        db_workflow_script = DBWorkflowScript(
+            workflow_id=workflow_id,
+            content=content,
+        )
+        await db_workflow_script.insert()
```

Comment:
Maybe we could check in the database first whether this exact workflow already exists, to avoid redundancy.

Reply:
> Maybe we could check in the database first whether this exact workflow already exists, to avoid redundancy.

This would degrade the DB, wouldn't it? We should not iterate over the content of every workflow and compare it with the content of the currently uploaded workflow.

Reply:
I could add a hash for a workflow and compare that, this way the database should not be too busy with it. I'll give it a try.

Reply:
Or just use `md5sum(content)` as the mongo `_id`?

Reply:
I would not alter the default handling of `_id` by assigning a user-defined variable to it. That `_id` is used internally for versioning from my understanding, and changing it manually may be problematic. Moreover, a single space in the same workflow would still lead to different hashes. But I am fine if we still go for the hash solution.

Reply:
OK, I have not worked intensively with MongoDB for many years, so pls disregard the `_id` hack idea.

A separate field with `hash(content.strip())` would still be nice though. Even if it does not catch all semantically equivalent workflows, it's still helpful I think to deduplicate.

Reply:
I have implemented a check for duplication. I use md5 to hash the file before saving it. The hash is stored in the database too and before saving it is checked that the same hash does not exist yet. I thought about removing trailing whitespace and similar stuff but decided against it.
I am not too sure about the benefit of this, but I think the content check does not hurt much, so I would keep it for now.